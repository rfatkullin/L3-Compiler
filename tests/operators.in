func main (args->[[char]]) -> int
	a = [int 100] -> [int];
	ind = 99 -> int;
	ch1, ch2 -> char;
	yes = tt -> bool;

	print("[Test]: Add operator");
		print(0 + 98);
		print(98 + 0);
		ch1 = 'a' + 1;
		ch2 = 1 + 'a';
		print(ch1);
		print(ch2);
	print("---------------------");

	print("[Test]: Sub operator");
		print(0 - 98);
		print(98 - 0);
		print('b' - 'a');
		ch1 = 'b' - 1;
		print(ch1);
	print("---------------------");

	print("[Test]: Multiply operator");
		print(1 2);
		print(13  5);
		print(0 100);
		print(10 8);
	print("---------------------");

	print("[Test]: Pow operator");
		print(2 ^ 10);
		print(0 ^ 2);
		print(1 ^ 100);
	print("---------------------");

	print("[Test]: Divide operator");
		print(2 / 2);
		print(5 / 2);
		print(1 / 100);
		print(56 / 8);
	print("---------------------");

	print("[Test]: Mod operator");
		print(1 % 2);
		print(13 % 5);
		print(10 % 100);
		print(0 % 8);
	print("---------------------");

	print("[Test]: Unary minus operator");
		print((-'a'));
		print((-'q'));
		print((-(1 + 2)));
	print("---------------------");

	print("[Test]: Logical not operator");
		print(yes);
		print(!yes);
		print(!(!yes));
	print("---------------------");

	print("[Test]: Get array element operator");
		a['a'] = 2;
		a[ind] = 1;
		print(ind);
		print(a[ind]);
		print(a['a']);
	print("---------------------");

	print("[Test]: Eq, not eq operators");
		print(1 == 1);
		print(1 != 1);
		print(1 == 2);
		print(1 != 2);
		print(' ');

		print(97 == 'a');
		print('a' == 97);
		print('a' != 98);
		print(98 != 'a');
		print(98 == 'a');
		print('a' == 98);
		print(' ');

		print('a' == 'a');
		print('a' != 'a');
		print('a' != 'b');
		print('a' == 'b');
		print(' ');

		print(tt == tt);
		print(tt != tt);
		print(tt == ff);
		print(tt != ff);
	print("---------------------");

	print("[Test]: Lss, gtr, lss_eq, gtr_eq operators");
		print('1');
		print(1 < 1);
		print(1 > 1);
		print(1 <= 1);
		print(1 >= 1);

		print('2');
		print(1 < 2);
		print(1 > 2);
		print(1 <= 2);
		print(1 >= 2);

		print('3');
		print(2 < 1);
		print(2 > 1);
		print(2 <= 1);
		print(2 >= 1);

		print('4');
		print('a' < 'a');
		print('a' > 'a');
		print('a' <= 'a');
		print('a' >= 'a');

		print('5');
		print('a' < 'b');
		print('a' > 'b');
		print('a' <= 'b');
		print('a' >= 'b');

		print('6');
		print('b' < 'a');
		print('b' > 'a');
		print('b' <= 'a');
		print('b' >= 'a');

		print('7');
		print('a' < 97);
		print('a' > 97);
		print('a' <= 97);
		print('a' >= 97);

		print('8');
		print(97 < 'a');
		print(97 > 'a');
		print(97 <= 'a');
		print(97 >= 'a');

		print('9');
		print('a' < 98);
		print('a' > 98);
		print('a' <= 98);
		print('a' >= 98);

		print("10");
		print(98 < 'a');
		print(98 > 'a');
		print(98 <= 'a');
		print(98 >= 'a');
	print("---------------------");

	print("[Test]: Logical and, or, xor operators");
		print(tt && tt);
		print(tt && ff);
		print(ff && tt);
		print(ff && ff);
		print(' ');

		print(tt || tt);
		print(tt || ff);
		print(ff || tt);
		print(ff || ff);
		print(' ');

		print(tt ^^ tt);
		print(tt ^^ ff);
		print(ff ^^ tt);
		print(ff ^^ ff);		
	print("---------------------");

	main = 0;
endfunc
