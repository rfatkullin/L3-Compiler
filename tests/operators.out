[Test]: Add operator
98
98
b
b
---------------------
[Test]: Sub operator
-98
98
1
a
---------------------
[Test]: Multiply operator
2
65
0
80
---------------------
[Test]: Pow operator
1024
0
1
---------------------
[Test]: Divide operator
1
2
0
7
---------------------
[Test]: Mod operator
1
3
10
0
---------------------
[Test]: Unary minus operator
-97
-113
-3
---------------------
[Test]: Logical not operator
True
False
True
---------------------
[Test]: Get array element operator
99
1
2
---------------------
[Test]: Eq, not eq operators
True
False
False
True
 
True
True
True
True
False
False
 
True
False
True
False
 
True
False
False
True
---------------------
[Test]: Lss, gtr, lss_eq, gtr_eq operators
1
False
False
True
True
2
True
False
True
False
3
False
True
False
True
4
False
False
True
True
5
True
False
True
False
6
False
True
False
True
7
False
False
True
True
8
False
False
True
True
9
True
False
True
False
10
False
True
False
True
---------------------
[Test]: Logical and, or, xor operators
True
False
False
False
 
True
True
True
False
 
False
True
True
False
---------------------
